### PR TITLE
System should not be added to multiple different formations of a kind

### PR DESCRIPTION
**Pull request by nyordanoff**
**Description**
This PR adds a formation constraint that forbids a System to be added to multiple different formations of a given type. The only exception is if the System itself has a type that is allowed to participate in many formations.

Changes proposed in this pull request:
- Rename existing formation constraint to has more generic name
- Add the new constraint
- Add logic to support `ANY` as value for a `resourceSubtype` when the constraint doesn't need to specify a concrete subtype 

**Pull Request status**

- [x] Implementation
- [x] Unit tests
- [ ] Integration tests
- [x] `chart/compass/values.yaml` is updated 
- [ ] Mocks are regenerated, using the automated script

### CODE REVIEW

File: tests/director/tests/formation_api_test.go [hunk 1]
```diff
@@ -695,6 +698,134 @@ func TestSubaccountInAtMostOneFormationOfType(t *testing.T) {
...
+	defer fixtures.CleanupFormationTemplate(t, ctx, certSecuredGraphQLClient, &formationTemplate)
+	require.NoError(t, err)
+	require.NotEmpty(t, formationTemplate.ID)
+	require.NotEmpty(t, formationTemplate.Name)
+
+	in := graphql.FormationConstraintInput{
+		Name:            "TestSystemInAtMostOneFormationOfType",
+		ConstraintType:  graphql.ConstraintTypePre,
+		TargetOperation: graphql.TargetOperationAssignFormation,
+		Operator:        "IsNotAssignedToAnyFormationOfType",
```
Comment (StanislavStefanov) [inline tests/director/tests/formation_api_test.go:731]: We should use the constant `IsNotAssignedToAnyFormationOfTypeOperator` for the operator name

File: tests/director/tests/formation_api_test.go [hunk 2]
```diff
@@ -695,6 +698,134 @@ func TestSubaccountInAtMostOneFormationOfType(t *testing.T) {
...
+	defer fixtures.UnassignFormationWithApplicationObjectType(t, ctx, certSecuredGraphQLClient, graphql.FormationInput{Name: firstFormation.Name}, app.ID, tenantId)
+	assignReq := fixtures.FixAssignFormationRequest(app.ID, string(graphql.FormationObjectTypeApplication), firstFormation.Name)
+	var assignFormation graphql.Formation
+	err = testctx.Tc.RunOperation(ctx, certSecuredGraphQLClient, assignReq, &assignFormation)
+	require.NoError(t, err)
+	require.Equal(t, firstFormation.Name, assignFormation.Name)
+
+	t.Logf("Should fail to assign first application to another formation with name: %s of the same kind", secondFormation.Name)
+	defer fixtures.UnassignFormationWithApplicationObjectType(t, ctx, certSecuredGraphQLClient, graphql.FormationInput{Name: secondFormation.Name}, app.ID, tenantId)
+	assignReq = fixtures.FixAssignFormationRequest(app.ID, string(graphql.FormationObjectTypeApplication), secondFormation.Name)
```
Comment (StanislavStefanov) [inline tests/director/tests/formation_api_test.go:795]: AssignFormationWithApplicationObjectTypeExpectError fixture could be used

File: tests/director/tests/formation_api_test.go [hunk 3]
```diff
@@ -695,6 +698,134 @@ func TestSubaccountInAtMostOneFormationOfType(t *testing.T) {
...
+	err = testctx.Tc.RunOperation(ctx, certSecuredGraphQLClient, assignReq, &assignFormation)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "Operator \"IsNotAssignedToAnyFormationOfType\" is not satisfied")
```
Comment (StanislavStefanov) [inline tests/director/tests/formation_api_test.go:798]: We should use the constant IsNotAssignedToAnyFormationOfTypeOperator for the operator name

File: tests/director/tests/formation_api_test.go [hunk 4]
```diff
@@ -695,6 +698,134 @@ func TestSubaccountInAtMostOneFormationOfType(t *testing.T) {
...
+	t.Log("Create second application with exception type")
+	exceptionTypeApp := graphql.ApplicationExt{} // needed so the 'defer' can be above the application creation
+	defer fixtures.CleanupApplication(t, ctx, certSecuredGraphQLClient, tenantId, &exceptionTypeApp)
+	exceptionTypeApp, err = fixtures.RegisterApplicationWithApplicationType(t, ctx, certSecuredGraphQLClient, "e2e-tests-exceptionType-app", conf.ApplicationTypeLabelKey, exceptionSystemType, tenantId)
+	require.NoError(t, err)
+	require.NotEmpty(t, exceptionTypeApp.ID)
+
+	t.Logf("Assign first application to first formation with name: %s", firstFormation.Name)
+	defer fixtures.UnassignFormationWithApplicationObjectType(t, ctx, certSecuredGraphQLClient, graphql.FormationInput{Name: firstFormation.Name}, app.ID, tenantId)
+	assignReq := fixtures.FixAssignFormationRequest(app.ID, string(graphql.FormationObjectTypeApplication), firstFormation.Name)
```
Comment (StanislavStefanov) [inline tests/director/tests/formation_api_test.go:787]: AssignFormationWithApplicationObjectType fixture could be used

File: tests/director/tests/formation_api_test.go [hunk 5]
```diff
@@ -695,6 +698,134 @@ func TestSubaccountInAtMostOneFormationOfType(t *testing.T) {
...
+
+	t.Logf("Should succeed assigning second application with exception type to another formation with name: %s of the same kind", secondFormation.Name)
+	defer fixtures.UnassignFormationWithApplicationObjectType(t, ctx, certSecuredGraphQLClient, graphql.FormationInput{Name: secondFormation.Name}, exceptionTypeApp.ID, tenantId)
+	assignReq = fixtures.FixAssignFormationRequest(exceptionTypeApp.ID, string(graphql.FormationObjectTypeApplication), secondFormation.Name)
+	err = testctx.Tc.RunOperation(ctx, certSecuredGraphQLClient, assignReq, &assignFormation)
+	require.NoError(t, err)
+	require.Equal(t, secondFormation.Name, assignFormation.Name)
+
+	t.Logf("Detaching constraint from formation template")
+	deleteRequest := fixtures.FixDetachConstraintFromFormationTemplateRequest(constraint.ID, formationTemplate.ID)
```
Comment (StanislavStefanov) [inline tests/director/tests/formation_api_test.go:815]: DetachConstraintFromFormationTemplate fixture could be used

**Review by StanislavStefanov — CHANGES_REQUESTED** [review]

File: tests/director/tests/formation_api_test.go (hunk 3 above)
Comment (nyordanoff) [reply tests/director/tests/formation_api_test.go:798]: Not applicable because the fixture that expects error doesn't check the actual error.

**Pull request merged**